train_gpt2.py

```diff
@@ -231,10 +232,22 @@ def __init__(self, B, T, process_rank, num_processes, split):
             print(f"found {len(shards)} shards for split {split}")
         self.reset()
 
+    def load_shard(self, filename):
+        shard = load_tokens(filename)
+        if self.split == "train":
+            # split tokens into documents using the <|endoftext|> token and shuffle
+            eot_positions = (torch.where(shard == enc.eot_token)[0] + 1).tolist()
+            documents = [shard[start:end] for start, end in zip([0] + eot_positions[:-1], eot_positions)]
+            np.random.shuffle(documents)
+            shard = torch.cat(documents) # concatenate the documents back together
+        return shard
+
     def reset(self):
         # state, init at shard zero
         self.current_shard = 0
-        self.tokens = load_tokens(self.shards[self.current_shard])
+        if self.split == "train":
+            np.random.shuffle(self.shards)
```

Comment:
I suggest you create a random generator, initialized differently on each rank, so the results can be more reproducible.
Can use: https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.shuffle.html#numpy.random.Generator.shuffle
E.g.:
```
rng = np.random.default_rng(self.process_rank + 1337)
...
rng.shuffle(self.shards)
```

Reply:
You're right, I've made a mistake!